File: py/LinkedList.py

#  delete key is buggy and need to verify edge for delete pos

class node:
    def __init__(self, data=None):
        self.data = data
        self.link = None


class LinkedList:
    def __init__(self, head=None):
        self.head = head
        self.non = 0

    def display(self):
        p = self.head
        print(f"No of elements in the list are:{self.non}")
        while p is not None:
            print(p.data, end=" ")
            p = p.link
        print("")
        if self.head == None:
            print("Empty!")

    def insert_front(self, data):
        temp = node(data)
        temp.link = self.head
        self.head = temp
        self.non = self.non + 1

    def insert_back(self, data):
        temp = node(data)
        if self.head == None:
            self.insert_front(data)
            return

        p = self.head
        while p.link is not None:
            p = p.link
        p.link = temp
        self.non = self.non + 1

    def insert_mid(self, data):
        if self.non == 0:
            self.insert_front(data)
            return

        if self.non == 1:
            if self.head.data > data:
                self.insert_front(data)
                return
            else:
                self.insert_back(data)
                return

        temp = node(data)
        p = self.head
        q = None
        while p != None and p.data < data:
            q = p
            p = p.link

        if p == None:
            q.link = temp
            self.non = self.non + 1
            return

        if q == None:
            self.insert_front(data)
            return
        q.link = temp
        temp.link = p
        self.non = self.non + 1

    def destroy(self):
        if self.non == 0:
            return

        if self.non == 1:
            del self.head
            return

        p = self.head
        q = None

        while p is not None:
            del q
            q = p
            p = p.link
        del q
        return

    def delete_front(self):
        if self.non == 0:
            return
        p = self.head
        self.head = p.link
        del p
        self.non = self.non-1

    def delete_back(self):
        if self.non == 0:
            return

        if self.non == 1:
            self.delete_front()
            return
        p = self.head
        q = None

        while p.link is not None:
            q = p
            p = p.link

        q.link = None
        del p
        self.non = self.non-1

    def delete_key(self, data):
        if self.non == 0:
            return

        if self.non == 1:
            if self.head.data == data:
                self.delete_front()
                return
            else:
                return

        if self.head.data == data:
            self.delete_front()
            return

        p = self.head
        q = None

        while p is not None and p.data != data:
            q = p
            p = p.link

        if p == None:
            return

        q.link = p.link
        del p
        self.non = self.non-1

    def delete_pos(self, pos):

        if self.non == 0:
            return

        if self.non == 1:
            if pos == 1:
                self.delete_front()
                return
            else:
                return

        p = self.head
        q = None

        if pos == 1:
            self.delete_front()
            return
        count = 1
        while count is not pos and p is not None:
            q = p
            p = p.link
            count = count+1

        if p == None:
            return

        q.link = p.link
        del p
        self.non = self.non - 1

    def search(self, data):
        if self.non == 0:
            return -1

        if self.non == 1:
            if self.head.data == data:
                return 1
            else:
                return -1

        if self.head.data == data:
            return 1

        p = self.head
        count = 1

        while p is not None and p.data != data:
            p = p.link
            count = count+1

        if p is None:
            return -1

        return count

    def duplicate(self, newlist):
        if self.non == 1:
            return

        if self.non == 1:
            newlist.head = node(self.head.data)
            newlist.non = newlist.non+1
            return

        p = self.head

        while p is not None:
            newlist.insert_back(p.data)
            p = p.link


list1 = LinkedList()
list1.head = None
list1.insert_front(1)
list1.insert_back(3)
list1.insert_back(4)
list1.insert_back(5)
list1.insert_back(6)
list1.insert_mid(0)


list1.display()

data = 4

print(f"data = {data} can be found at pos {list1.search(data)}")

list2 = LinkedList()

list1.duplicate(list2)

list2.display()


list1.destroy()
del list1

File: cpp/LinkedList.cpp
#include <iostream>

// void display(LL *);
// node *create_node(int);
// void insert_front(LL *, int);
// void insert_back(LL *, int);
// void destroy(LL *);
// void insert_mid(LL *, int);
// void delete_back(LL *);
// void delete_front(LL *);
// void delete_key(LL *, int);
// void delete_pos(LL *, int);
// int search(LL *, int);
// LL *duplicate(LL *);
/*
sort
merge (into sorted)
*/

using namespace std;

class node
{
public:
  int data;
  struct node *link;
  node(int x = 0, struct node *y = NULL)
  {
    data = x;
    link = y;
  }
};

typedef struct node node;

class LinkedList
{
public:
  int non;
  node *head;

  LinkedList()
  {
    non = 0;
    head = NULL;
  }

  void insert_front(int data)
  {
    node *temp = new node(data);

    if (non == 0)
    {
      head = temp;
      non++;
      return;
    }

    temp->link = head;
    head = temp;
    non++;
    return;
  }
  void display()
  {
    cout << "No of nodes in the list: " << non << endl;
    if (non == 0)
    {
      cout << "Empty!";
      return;
    }

    if (non == 1)
    {
      cout << head->data << " " << endl;
      return;
    }

    node *p = head, *q = NULL;

    while (p != NULL)
    {
      cout << p->data << " ";
      p = p->link;
    }
    cout << endl;
  }
  void insert_back(int data)
  {
    if (non == 0)
    {
      insert_front(data);
      return;
    }

    node *temp = new node(data);

    if (non == 1)
    {
      head->link = temp;
      non++;
      return;
    }

    node *p = head;

    while (p->link != NULL)
    {
      p = p->link;
    }

    p->link = temp;
    non++;
  }

  void insert_mid(int data)
  {
    if (non == 0)
    {
      insert_front(data);
      return;
    }

    if (non == 1)
    {
      if (data > head->data)
      {
        insert_back(data);
        return;
      }
      insert_front(data);
      return;
    }

    if (head->data > data)
    {
      insert_front(data);
      return;
    }

    node *temp = new node, *p = head, *q = NULL;

    while (p != NULL && data > p->data)
    {
      q = p;
      p = p->link;
    }

    if (p == NULL)
    {
      insert_back(data);
      return;
    }

    q->link = temp;
    temp->link = p;
    non++;
  }

  void destroy()
  {
    if (non == 0)
    {
      return;
    }
    if (non == 1)
    {
      delete head;
      return;
    }

    node *p = head, *q = NULL;

    while (p != NULL)
    {
      q = p;
      p = p->link;
      delete q;
    }
    non = 0;
  }

  void delete_front()
  {
    if (non == 0)
    {
      return;
    }

    node *p = head;
    head = head->link;
    delete p;
    non--;
  }

  void delete_back()
  {
    if (non == 0)
    {
      return;
    }
    if (non == 1)
    {
      delete_front();
      return;
    }

    node *p = head;

    while (p->link->link != NULL)
    {
      p = p->link;
    }
    delete p->link;
    p->link = NULL;
    non--;
  }

  void delete_key(int key)
  {
    if (non == 0)
      return;

    if (non == 1 && head->data == key)
    {

      delete_back();
      return;
    }
    else
      return;

    if (head->data = key)
    {
      delete_front();
      return;
    }

    node *q = NULL, *p = head;
    cout << "hieeeeeeeee" << endl;

    while (p->data != key)
    {
      q = p;
      p = p->link;
    }
    q->link = p->link;
    delete p;
    non--;
  }
};

typedef struct LinkedList LL;

int main()
{
  LL *list1 = new LL;
  list1->insert_mid(7);
  list1->insert_mid(4);
  list1->insert_mid(1);
  list1->insert_mid(100);

  list1->delete_front();
  list1->delete_back();
  list1->delete_key(4);
  list1->display();

  list1->destroy();
  delete list1;
}
